### PR TITLE
feat(ratelimit): Add Attributes support to rate limit middleware

### PR DESCRIPTION
**Pull request by nickvergessen**
## Summary

* Sibling to #36928

### Previously (and still supported)
```php
	/**
	 * @UserRateThrottle(limit=20, period=200)
	 * @AnonRateThrottle(limit=10, period=100)
	 */
	public function testMethodWithAnnotation() {
```

### New syntax
```php
	#[UserRateLimit(limit: 20, period: 200)]
	#[AnonRateLimit(limit: 10, period: 100)]
	public function testMethodWithAttribute() {
```

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required https://github.com/nextcloud/documentation/pull/10100
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by kesselb — COMMENTED** [review]
Nice :+1:

File: lib/public/AppFramework/Http/Attribute/ARateLimit.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace OCP\AppFramework\Http\Attribute;
+
+use Attribute;
```
Comment (come-nc) [inline lib/public/AppFramework/Http/Attribute/ARateLimit.php:28]:
```suggestion
```

**Review by come-nc — APPROVED** [review]

**Comment by ChristophWurst** [comment]
> Previously (and still supported)

I would prefer to deprecate the old system right away so we can move everyone to attribute in the next few years and drop the legacy system as soon as possible.

**Comment by nickvergessen** [comment]
> I would prefer to deprecate the old system right away so we can move everyone to attribute in the next few years and drop the legacy system as soon as possible.

Yeah that was my plan. The problem is I didn't find a way (without totally spaming out of control on debug) to notify devs about the deprecation. So yeah a section in the "Changes for 27" will be added for BruteForceProtection and RateLimit later on

File: lib/public/AppFramework/Http/Attribute/ARateLimit.php (hunk 1 above)
Comment (nickvergessen) [reply lib/public/AppFramework/Http/Attribute/ARateLimit.php:28]: Fixed

**Review by vitormattos — APPROVED** [review]

File: lib/public/AppFramework/Http/Attribute/ARateLimit.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * Attribute for controller methods that want to limit the times a logged-in
+ * user can call the endpoint in a given time period.
+ *
+ * @since 27.0.0
+ */
+abstract class ARateLimit {
+	/**
+	 * @since 27.0.0
+	 */
```
[the remark this answers is not in the document]
Comment (ChristophWurst) [reply lib/public/AppFramework/Http/Attribute/ARateLimit.php:37]: `@since` is required for all functions, classes and methods in the `OCP\` namespace

**Pull request merged**